### PR TITLE
[EagerAppCDS] EagerAppCDS implementation

### PR DESCRIPTION
**Pull request by zhengxiaolinX**
Summary: An experimental feature, an eager version of AppCDS. Documents and Wiki page are to be filled.

Test Plan: tested internally; tests are in another patch.

Reviewed-by: yuleil, lingjun-cg

Issue: #345

### CODE REVIEW

File: src/hotspot/share/classfile/classLoader.cpp [hunk 1]
```diff
@@ -501,20 +504,8 @@ ClassFileStream* ClassPathImageEntry::open_stream(const char* name, TRAPS) {
     if (pkg_name != NULL) {
       if (!Universe::is_module_initialized()) {
         location = (*JImageFindResource)(_jimage, JAVA_BASE_NAME, get_jimage_version_string(), name, &size);
-#if INCLUDE_CDS
```
Comment (yuleil) [inline src/hotspot/share/classfile/classLoader.cpp:504]: 这段为啥删了？

**Review by lingjun-cg — APPROVED** [review]

File: src/hotspot/share/classfile/systemDictionaryShared.cpp [hunk 2]
```diff
@@ -654,25 +900,45 @@ InstanceKlass* SystemDictionaryShared::acquire_class_for_current_thread(
                                                   cfs, THREAD);
   if (shared_klass == NULL || HAS_PENDING_EXCEPTION) {
     // TODO: clean up <ik> so it can be used again
+    if (EagerAppCDS) {
+      ResourceMark rm;
+      log_trace(class, eagerappcds) ("[CDS load class] Failed: load_shared_class check super/interface/check_shared_class_file_load_hook failed: class %s; in class loader %p (%x)", ik->name()->as_C_string(),
```
[the remark this answers is not in the document]
Comment (zhengxiaolinX) [reply src/hotspot/share/classfile/systemDictionaryShared.cpp]: Nice catch. Done.

File: src/hotspot/share/classfile/systemDictionary.cpp [hunk 3]
```diff
@@ -1398,8 +1477,77 @@ void SystemDictionary::clear_invoke_method_table() {
     }
   }
 }
+
+bool SystemDictionary::should_not_dump_class(InstanceKlass* k) {
```
Comment (yuleil) [inline src/hotspot/share/classfile/systemDictionary.cpp:1481]: Add log here to help us understand why many classes cannot be dumped by CDS

Comment (zhengxiaolinX) [reply src/hotspot/share/classfile/systemDictionary.cpp:1481]: Done.

**Review by yuleil — APPROVED** [review]

**Pull request merged**